expr/func_aspercent.go

```diff
@@ -0,0 +1,251 @@
+package expr
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sort"
+	"strings"
+
+	"github.com/grafana/metrictank/api/models"
+	schema "gopkg.in/raintank/schema.v1"
+)
+
+type FuncAsPercent struct {
+	in          GraphiteFunc
+	totalFloat  float64
+	totalSeries GraphiteFunc
+	nodes       []expr
+}
+
+func NewAsPercent() GraphiteFunc {
+	return &FuncAsPercent{totalFloat: math.NaN()}
+}
+
+func (s *FuncAsPercent) Signature() ([]Arg, []Arg) {
+	return []Arg{
+		ArgSeriesList{val: &s.in},
+		ArgIn{
+			key: "total",
+			opt: true,
+			args: []Arg{
+				ArgFloat{val: &s.totalFloat},
+				ArgSeriesList{val: &s.totalSeries},
+			},
+		},
+		ArgStringsOrInts{val: &s.nodes, opt: true, key: "nodes"},
+	}, []Arg{ArgSeriesList{}}
+}
+
+func (s *FuncAsPercent) Context(context Context) Context {
+	return context
+}
+
+func (s *FuncAsPercent) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
+	series, err := s.in.Exec(cache)
+	if err != nil {
+		return nil, err
+	}
+
+	var outSeries []models.Series
+	defer func() { cache[Req{}] = append(cache[Req{}], outSeries...) }()
+	var totals []models.Series
+	if s.totalSeries != nil {
+		totals, err = s.totalSeries.Exec(cache)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
If totals is zero length, it is not an error. Instead, it will return a bunch of empty series with names something like `asPercent(foo.*,MISSING)`

Reply:
Ohhh I see what you mean. Because the totalsSeries has to be either `== 1` or `== len(series)`

But the error will get triggered at a later check. Plus, if the series is empty as well, then it should trigger an error technically

Reply:
Generally I think it's better to return the error early when we can. So when reasoning about the code we have fewer special cases that complicate things further down.

Reply:
Ok, I moved up the validation like you suggested, so the check for proper length of totals is right after it. (Gonna push soon just looking at one more thing)

Reply:
@stivenbb ^^ why can't we just return an error here if `len(totals)` is 0 ? this case still slips through into the `if s.nodes != ni` case for example

Reply:
When `nodes != nil` and `len(totals) == 0`, it shouldn't fail. Instead, all targets will be `asPercent(foo,MISSING)`

Reply:
aha! ok thx